### PR TITLE
RF: Add impurity variable importance

### PR DESCRIPTION
**Pull request by orhankislal**
JIRA: MADLIB-1205

This commit makes the following changes:
- Add impurity variable importance for random forests.
- Rename current cat_var_importance and con_var_importance measurements to
oob_cat_var_importance and oob_con_var_importance.

New impurity measurement is provided as impurity_var_importance, and supports
grouping. It combines the importance values for both categorical and
continuous features into a single array.

Co-authored-by: Rahul Iyer <riyer@pivotal.io>
Co-authored-by: Jingyi Mei <jmei@pivotal.io>
Co-authored-by: Arvind Sridhar <asridhar@pivotal.io>
Co-authored-by: Nandish Jayaram <njayaram@apache.org>

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/538/

File: src/ports/postgres/modules/recursive_partitioning/random_forest.py_in [hunk 1]
```diff
@@ -1333,42 +1368,69 @@ def _create_group_table(
 # -------------------------------------------------------------------------
 
 
-def _create_empty_result_table(schema_madlib, output_table_name):
+def _create_empty_result_table(schema_madlib, output_table_name, importance):
     """Create the result table for all trees in the forest"""
+    impurity_var_imp_str = """, impurity_var_importance double precision[]);""" if importance else ");"
```
Comment (iyerr3) [inline src/ports/postgres/modules/recursive_partitioning/random_forest.py_in]: Do we need the triple quotes? Also if `);` is common in both cases then we can move that to the query directly?

File: src/ports/postgres/modules/recursive_partitioning/random_forest.sql_in [hunk 2]
```diff
@@ -249,9 +250,13 @@ forest_train(training_table_name,
       </tr>
       <tr>
         <th>tree</th>
-        <td>BYTEA8. Trained tree model stored in binary 
+        <td>BYTEA8. Trained tree model stored in binary
         format (not human readable).</td>
       </tr>
+      <tr>
+        <th>impurity_var_importance</th>
+        <td>DOUBLE PRECISION[]. The gini impurity importance score for the tree.</td>
```
Comment (jingyimei) [inline src/ports/postgres/modules/recursive_partitioning/random_forest.sql_in]: Also mention the order corresponds to the order of the variables as found in    'cat_features' followed by the order of the 'con_features' in summary table

File: src/ports/postgres/modules/recursive_partitioning/random_forest.py_in [hunk 3]
```diff
@@ -1333,42 +1368,69 @@ def _create_group_table(
...
 
 
 def _insert_into_result_table(schema_madlib, tree_states, output_table_name,
-                              grp_key_to_grp_cols, sample_id):
+                              grp_key_to_grp_cols, sample_id, importance, grouping_cols):
     """Insert one tree to result table"""
+
+    impurity_var_imp_str = ''
+    importance_query = ''
+    importance_results = ''
```
Comment (iyerr3) [inline src/ports/postgres/modules/recursive_partitioning/random_forest.py_in]: IMO these variables in else clause is easier to read. Also `importance_results` is limited to the `if` block and does not require to be defined outside that scope.

File: src/ports/postgres/modules/recursive_partitioning/random_forest.py_in [hunk 4]
```diff
@@ -1291,38 +1300,64 @@ def _create_group_table(
...
                 {grouping_cols_str}
-                grp_finished as success,
+                grp_finished AS success,
                 cat_n_levels,
                 cat_levels_in_text,
-                oob_error,
-                cat_var_importance,
-                con_var_importance
+                oob_error
+                {cat_var_importance_str}
```
Comment (jingyimei) [inline src/ports/postgres/modules/recursive_partitioning/random_forest.py_in]: We can make it consistent with oob_*

File: src/ports/postgres/modules/recursive_partitioning/random_forest.py_in [hunk 5]
```diff
@@ -1291,38 +1300,64 @@ def _create_group_table(
...
+                    unnest($2) AS grp_finished
             ) tree_terminated
             USING (grp_key)
             JOIN
                 {cat_features_info_table}
             USING (gid)
-            LEFT OUTER JOIN
-                {importance_table}
-            USING (gid)
+            {left_join_importance_table_str}
```
Comment (jingyimei) [inline src/ports/postgres/modules/recursive_partitioning/random_forest.py_in]: We can name it oob_importance_table_* to explicitly distinguish it from impurity importance table

File: src/modules/recursive_partitioning/DT_impl.hpp [hunk 6]
```diff
@@ -1512,6 +1512,9 @@ DecisionTree<Container>::computeVariableImportance(
                     uint64_t maj_count = getMajorityCount(node_index);
                     uint64_t min_count =  node_count - maj_count;
 
+                    if (min_count == 0) {
```
[the remark this answers is not in the document]
Comment (iyerr3) [reply src/modules/recursive_partitioning/DT_impl.hpp]: I've pushed a commit to master that fixes the need to do this hack. Please rebase on master and remove these lines.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/561/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/563/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/572/

**Pull request closed without merging**